build.go

```diff
@@ -0,0 +1,163 @@
+}
+
+func Build(
+	dependencyService DependencyService,
+	calculator Calculator,
+	sbomGenerator SBOMGenerator,
+	logger scribe.Emitter,
+	clock chronos.Clock,
+) packit.BuildFunc {
+	return func(context packit.BuildContext) (packit.BuildResult, error) {
+		logger.Title("%s %s", context.BuildpackInfo.Name, context.BuildpackInfo.Version)
+
+		planner := draft.NewPlanner()
+
+		logger.Process("Resolving VSCodium version")
+		entry, sortedEntries := planner.Resolve("vscodium", context.Plan.Entries, []interface{}{
+			"BP_VSCODIUM_VERSION",
+			"buildpack.toml",
+		})
+		logger.Candidates(sortedEntries)
+
+		entryVersion, _ := entry.Metadata["version"].(string)
+		dependency, err := dependencyService.Resolve(filepath.Join(context.CNBPath, "buildpack.toml"), entry.Name, entryVersion, context.Stack)
+		if err != nil {
+			return packit.BuildResult{}, err
+		}
+
+		logger.SelectedDependency(entry, dependency, clock.Now())
+
+		layer, err := context.Layers.Get(VSCODIUM)
+		if err != nil {
+			return packit.BuildResult{}, err
+		}
+
+		bom := dependencyService.GenerateBillOfMaterials(dependency)
+		launch, build := planner.MergeLayerTypes(VSCODIUM, context.Plan.Entries)
+
+		var buildMetadata packit.BuildMetadata
+		if build {
+			buildMetadata.BOM = bom
+		}
+
+		var launchMetadata packit.LaunchMetadata
+		if launch {
+			command := "codium-server"
+			args := []string{
+				"--server-base-path", "${RENKU_BASE_URL_PATH}",
+				"--host", "${RENKU_SESSION_IP}",
+				"--port", "${RENKU_SESSION_PORT}",
+				"--extensions-dir", "${RENKU_MOUNT_DIR}/.vscode/extensions",
+				"--server-data-dir", "${RENKU_MOUNT_DIR}/.vscode",
+				"--without-connection-token",
+				"--accept-server-license-terms",
+				"--telemetry-level", "off",
+				"--default-folder", "${RENKU_WORKING_DIR}",
+			}
+			launchMetadata.Processes = []packit.Process{
+				{
+					Type:    "web",
+					Command: command,
+					Args:    args,
+					Default: true,
+					Direct:  false,
+				},
+			}
+			launchMetadata.BOM = bom
+
+		}
+
+		logger.Process("Executing build process")
+
+		// TODO: check layer caching so we don't reinstall if already installed
+		layer, err = layer.Reset()
+		if err != nil {
+			return packit.BuildResult{}, err
+		}
+
+		layer.Launch, layer.Build = launch, build
+
+		logger.Subprocess("Installing VSCodium %s", dependency.Version)
+		duration, err := clock.Measure(func() error {
+			return dependencyService.Deliver(dependency, context.CNBPath, layer.Path, context.Platform.Path)
+		})
+		if err != nil {
+			return packit.BuildResult{}, err
+		}
+
+		layer.Metadata = map[string]interface{}{
+			DepKey: dependency.Checksum,
+		}
+
+		logger.Action("Completed in %s", duration.Round(time.Millisecond))
+		logger.Break()
+
+		layer.SharedEnv.Append("PATH", filepath.Join(layer.Path, "bin"), string(os.PathListSeparator))
+
+		layer.ExecD = []string{filepath.Join(context.CNBPath, "bin", "install-extensions")}
```

Comment:
question: So this is launched as an init step before the vscode server is launched? When the vscode server starts?

Reply:
yes, this runs on session start before the server is launched.

actually I wanted to change this and bake the extensions into the image. But you can't write to the workdir from the buildpack and you can't write to the buildpack layer when running the image. and vscode only allows one extension dir. So if it was baked into the image, a user could not install additional extensions through the UI

Reply:
No that is fine. This was just to confirm that my understanding is correct.